### PR TITLE
feat: Firebase setup - Added Android native method for requesting permissions and fetching device tokens.

### PR DESCRIPTION
**Pull request by vijaygojiya**

### CODE REVIEW

File: android/src/main/java/com/candlefinance/push/PushModule.kt [hunk 1]
```diff
@@ -19,6 +29,61 @@ class PushModule(reactContext: ReactApplicationContext) :
...
+          }
+        }
+
+    } else {
+       promise.resolve(true)
+      }
+  }
+
+  @ReactMethod
+  fun getToken(promise: Promise) {
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/PushModule.kt]: #req to match iOS 
```suggestion
  fun registerForToken(promise: Promise) {
```

File: android/src/main/java/com/candlefinance/push/PushModule.kt [hunk 2]
```diff
@@ -19,6 +29,61 @@ class PushModule(reactContext: ReactApplicationContext) :
...
+            val currentRequestCode = 83834
+
+            val listener = PermissionListener { requestCode: Int, _: Array<String>, grantResults: IntArray ->
+              if (requestCode == currentRequestCode) {
+                val permissionStatus = if (grantResults.isNotEmpty()) grantResults[0] else PackageManager.PERMISSION_DENIED
+                promise.resolve(permissionStatus== PackageManager.PERMISSION_GRANTED)
+                return@PermissionListener true
+              }
+              return@PermissionListener false
+            }
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/PushModule.kt]:
```suggestion
            val listener = PermissionListener { requestCode: Int, _: Array<String>, grantResults: IntArray ->
    if (requestCode == currentRequestCode) {
        val isPermissionGranted = grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED
        promise.resolve(isPermissionGranted)
        return@PermissionListener true
    }
    return@PermissionListener false
}
```

Comment (gtokman) [reply android/src/main/java/com/candlefinance/push/PushModule.kt]: #nth a bit more concise

File: example/src/App.tsx [hunk 3]
```diff
@@ -70,6 +70,19 @@ export default function App() {
...
+      <Button
+        title="Get FCM Token"
+        onPress={() =>
+          push
+            .getToken()
+            .then((token) => {
+              console.log('Token :', token);
+            })
+            .catch((e) => {
+              console.log('Error while fetching deivce FCM token', e);
```
Comment (gtokman) [inline example/src/App.tsx]: #req can remove this and use the button above

File: src/index.tsx [hunk 4]
```diff
@@ -48,19 +48,19 @@ class Push {
...
-    if (Platform.OS === 'ios') {
-      const result = await this.module.requestPermissions();
-      return result;
-    }
-    return false;
+    // if (Platform.OS === 'ios') {
+    const result = await this.module.requestPermissions();
+    return result;
+    // }
+    // return false;
```
Comment (gtokman) [inline src/index.tsx]: #req delete these

File: src/index.tsx [hunk 5]
```diff
@@ -70,6 +70,13 @@ class Push {
     return false;
   }
 
+  public async getToken(): Promise<boolean> {
+    if (Platform.OS === 'android') {
+      return this.module.getToken();
+    }
+    return false;
+  }
+
```
Comment (gtokman) [inline src/index.tsx]: #req lets use the function `registerForToken` and remove this

**Review by gtokman — CHANGES_REQUESTED** [review]
Nice! thanks for this, few comments.

Last part is the listeners: https://reactnative.dev/docs/native-modules-android#sending-events-to-javascript

**Review by gtokman — APPROVED** [review]
Nice 🚀

**Comment by gtokman** [comment]
#1 adds the first part; now, all that is left is the emitter part (https://reactnative.dev/docs/native-modules-android#sending-events-to-javascript)

@vijaygojiya Let me know if you want to tackle that. I'm imagining something like this:
```kotlin
class PushPackage : ReactPackage {
  
  data class Action (val type: String, val payload: Map<String, Any?>)

  private var isInitialized = false
  private val queue = mutableListOf<Action>()
  private var module: PushModule? = null

  override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> {
    module = PushModule(reactContext)

    module?.onInitialized = {
      isInitialized = true
      queue.forEach {
        module?.dispatch(it.type, it.payload)
      }
      queue.clear()
    }
    
    return listOf(module as NativeModule)

  }

  override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> {
    return emptyList()
  }

  fun dispatch(action: String, payload: Map<String, Any?>) {
    if (isInitialized) {
      module?.dispatch(action, payload)
    } else {
      queue.add(Action(action, payload))
    }
  }
}
```
// in PushModule

```kotlin
 companion object {
    const val notificationReceived = "notificationReceived"
    const val deviceTokenReceived = "deviceTokenReceived"
    const val errorReceived = "errorReceived"
  }

  override fun getConstants() = mapOf(
         ...
  )
  
 // Example method
  // See https://reactnative.dev/docs/native-modules-android
  @ReactMethod
  fun addListener(eventName: String?) {
    // Keep: Required for RN built in Event Emitter Calls.
  }

  @ReactMethod
  fun removeListeners(count: Int?) {
    // Keep: Required for RN built in Event Emitter Calls.
  }

  fun dispatch(action: String, payload: Map<String, Any?>) {
    val map = mapOf(
        "type" to action,
        "payload" to payload
      )
      val event: WritableMap = Arguments.makeNativeMap(map)
      reactApplicationContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
        .emit("...", event)
  }
  ```

**Comment by vijaygojiya** [comment]
> #1 adds the first part; now, all that is left is the emitter part (https://reactnative.dev/docs/native-modules-android#sending-events-to-javascript)
> 
> @vijaygojiya Let me know if you want to tackle that. I'm imagining something like this:
> ```kotlin
> class PushPackage : ReactPackage {
>   
>   data class Action (val type: String, val payload: Map<String, Any?>)
> 
>   private var isInitialized = false
>   private val queue = mutableListOf<Action>()
>   private var module: PushModule? = null
> 
>   override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> {
>     module = PushModule(reactContext)
> 
>     module?.onInitialized = {
>       isInitialized = true
>       queue.forEach {
>         module?.dispatch(it.type, it.payload)
>       }
>       queue.clear()
>     }
>     
>     return listOf(module as NativeModule)
> 
>   }
> 
>   override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> {
>     return emptyList()
>   }
> 
>   fun dispatch(action: String, payload: Map<String, Any?>) {
>     if (isInitialized) {
>       module?.dispatch(action, payload)
>     } else {
>       queue.add(Action(action, payload))
>     }
>   }
> }
> ```
> // in PushModule
> 
> ```kotlin
>  companion object {
>     const val notificationReceived = "notificationReceived"
>     const val deviceTokenReceived = "deviceTokenReceived"
>     const val errorReceived = "errorReceived"
>   }
> 
>   override fun getConstants() = mapOf(
>          ...
>   )
>   
>  // Example method
>   // See https://reactnative.dev/docs/native-modules-android
>   @ReactMethod
>   fun addListener(eventName: String?) {
>     // Keep: Required for RN built in Event Emitter Calls.
>   }
> 
>   @ReactMethod
>   fun removeListeners(count: Int?) {
>     // Keep: Required for RN built in Event Emitter Calls.
>   }
> 
>   fun dispatch(action: String, payload: Map<String, Any?>) {
>     val map = mapOf(
>         "type" to action,
>         "payload" to payload
>       )
>       val event: WritableMap = Arguments.makeNativeMap(map)
>       reactApplicationContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
>         .emit("...", event)
>   }
>   ```
> 
> 
> 

Yes, I will take care of this today. Thanks

**Comment by vijaygojiya** [comment]
@gtokman Could you please review my implementation for the events emitter?. also need help on understanding event queue and dispatching event actions.

File: android/src/main/java/com/candlefinance/push/MyFirebaseMessagingService.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package com.candlefinance.push
+
+import android.util.Log
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+
+class MyFirebaseMessagingService : FirebaseMessagingService() {
+
+  // fun onSendError(messageId: String?, sendError: Exception?) {
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/MyFirebaseMessagingService.kt]: #req I'd loose the `My` prefix

File: android/src/main/java/com/candlefinance/push/MyFirebaseMessagingService.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  // //TODO handler error
+  //     Log.d(TAG,"on error")
+  // }
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/MyFirebaseMessagingService.kt]: #req remove this

File: android/src/main/java/com/candlefinance/push/NotificationUtils.kt [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import androidx.core.app.NotificationCompat
+import com.google.android.gms.common.internal.ResourceUtils
+import java.util.Locale
+
+
+object NotificationUtils {
+
+  private  val CHANNEL_ID = "RNFireDefaultChannelID"
+  private  val CHANNEL_NAME = "Firebase Default"
+  private val NOTIFICATION_ID = 123321
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/NotificationUtils.kt]: #? do we need to inject these from the JS side? If so, we might need to add an init function. I have ideas but never done Android push.

File: android/src/main/java/com/candlefinance/push/NotificationUtils.kt [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        CHANNEL_ID,
+        CHANNEL_NAME,
+        NotificationManager.IMPORTANCE_DEFAULT
+      )
+      channel.description = "All by default notification channel for all firebase notification"
+      notificationManager.createNotificationChannel(channel)
+    }
+
+  }
+  fun sendNotification(context: Context, title: String, message: String) {
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/NotificationUtils.kt:43]: #req add new line

File: android/src/main/java/com/candlefinance/push/NotificationUtils.kt [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+    val builder = NotificationCompat.Builder(context, CHANNEL_ID)
+      .setSmallIcon(getResourceIdByName("ic_launcher","mipmap"))
+      .setContentTitle(title)
+      .setContentText(message)
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/NotificationUtils.kt]: #? do we need to inject this, too? Is this like the app icon shown in the push?

File: src/index.tsx [hunk 11]
```diff
@@ -83,35 +75,29 @@ class Push {
...
     event: T,
     callback: NotificationCallbacks[T]
   ): void {
-    if (Platform.OS === 'ios') {
-      this.bridge?.addListener(event, callback);
-    }
+    this.bridge?.addListener(event, callback);
   }
 
   public removeListener<T extends keyof NotificationCallbacks>(event: T): void {
```
Comment (gtokman) [inline src/index.tsx:100]: #req remove ios check

File: android/src/main/java/com/candlefinance/push/MyFirebaseMessagingService.kt [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+  override fun onMessageReceived(remoteMessage: RemoteMessage) {
+    val title = remoteMessage.notification?.title.toString();
+    val body = remoteMessage.notification?.body.toString();
+
+    // TODO remove after testing 
+    RNEventEmitter.sendEvent("deviceTokenReceived",title)
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/MyFirebaseMessagingService.kt]: #? can you add a companion object

```
 companion object {
    const val notificationReceived = "notificationReceived"
    const val deviceTokenReceived = "deviceTokenReceived"
    const val errorReceived = "errorReceived"
  }
  
```

File: android/src/main/java/com/candlefinance/push/PushModule.kt [hunk 13]
```diff
@@ -12,13 +21,93 @@ class PushModule(reactContext: ReactApplicationContext) :
...
-    promise.resolve(a * b)
+  fun addListener(type: String?) {
+    // Keep: Required for RN built in Event Emitter Calls.
   }
 
+  @ReactMethod
+  fun removeListeners(type: Int?) {
+    // Keep: Required for RN built in Event Emitter Calls.
+  }
   companion object {
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/PushModule.kt]: #req new line

File: android/src/main/java/com/candlefinance/push/RNEventEmitter.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+
+  static void sendEvent(String eventName, String eventMap) {
+    var reactContext = ContextHolder.getApplicationContext();
+    try {
+
+      if (reactContext == null) {
+        return;
+      }
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/RNEventEmitter.java]: #nth log error

File: android/src/main/java/com/candlefinance/push/PushModule.kt [hunk 15]
```diff
@@ -12,13 +21,93 @@ class PushModule(reactContext: ReactApplicationContext) :
...
+        }
+
+    } else {
+       promise.resolve(true)
+      }
+  }
+
+  @ReactMethod
+  fun registerForToken(promise: Promise) {
+    FirebaseMessaging.getInstance().token.addOnCompleteListener(OnCompleteListener { task ->
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/PushModule.kt]: #req can you add the last method and remove the ios check in the index.ts

```
  @ReactMethod
  fun isRegisteredForRemoteNotifications(promise: Promise) {
    val token = FirebaseMessaging.getInstance().token
    promise.resolve(token != null)
  }
  ```

**Review by gtokman — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/candlefinance/push/RNEventEmitter.java [hunk 16]
```diff
@@ -0,0 +1,33 @@
+package com.candlefinance.push;
+
+import android.util.Log;
+
+import com.facebook.react.modules.core.DeviceEventManagerModule;
+
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/RNEventEmitter.java]: #? can we make this Kotlin

File: android/src/main/java/com/candlefinance/push/ContextHolder.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
+package com.candlefinance.push;
+import com.facebook.react.bridge.ReactContext;
+
+public class ContextHolder {
```
Comment (gtokman) [inline android/src/main/java/com/candlefinance/push/ContextHolder.java]: #req make this Kotlin

**Comment by gtokman** [comment]
> @gtokman Could you please review my implementation for the events emitter?. also need help on understanding event queue and dispatching event actions.

We probably don't need the queue, but you are good with the RNEmitter singleton you made regarding the dispatching events. After responding to my feedback, I will prob have to make a demo app on Firebase and see if this works.

File: android/src/main/java/com/candlefinance/push/NotificationUtils.kt (hunk 8 above)
Comment (vijaygojiya) [reply android/src/main/java/com/candlefinance/push/NotificationUtils.kt]: - Will do more R& D on it and improve.

File: android/src/main/java/com/candlefinance/push/NotificationUtils.kt (hunk 10 above)
Comment (vijaygojiya) [reply android/src/main/java/com/candlefinance/push/NotificationUtils.kt]: Essentially, we now require an icon when sending notifications, and I've also introduced a feature to easily change this icon. Developers can customize the notification icon by adding a new asset to the 'drawable' folder with the exact name ```ic_default_notification```.

**Pull request merged**